cocotb/bfmgen.py

```diff
@@ -0,0 +1,364 @@
+        with open(filename_c, "w") as out_c:
+            out_sv.write("//***************************************************************************\n")
+            out_sv.write("//* BFMs file for CocoTB. \n")
+            out_sv.write("//* Note: This file is generated. Do Not Edit\n")
+            out_sv.write("//***************************************************************************\n")
+
+            out_c.write("//***************************************************************************\n")
+            out_c.write("//* BFMs DPI interface file for CocoTB. \n")
+            out_c.write("//* Note: This file is generated. Do Not Edit\n")
+            out_c.write("//***************************************************************************\n")
+            out_c.write("#include <stdio.h>\n")
+            out_c.write("#ifdef __cplusplus\n")
+            out_c.write("extern \"C\" {\n")
+            out_c.write("#endif\n")
+            out_c.write("\n")
+            out_c.write("#include \"svdpi.h\"\n")
+            out_c.write("typedef void (*cocotb_bfm_notify_f)(void *);\n")
+            out_c.write("int cocotb_bfm_register(const char *, const char *, const char *, cocotb_bfm_notify_f, void *);\n")
+
+            for t in inst.bfm_type_info_m.keys():
+                info = inst.bfm_type_info_m[t]
+        
+                if cocotb.bfm_vlog not in info.hdl.keys():
+                    raise Exception("BFM \"" + t.__name__ + "\" does not support Verilog")
+        
+                with open(info.hdl[cocotb.bfm_sv], "r") as template_f:
+                    template = template_f.read()
+        
+                bfm_name = os.path.basename(info.hdl[cocotb.bfm_sv])
+    
+                if bfm_name.find('.') != -1:
+                    bfm_name = os.path.splitext(bfm_name)[0]
+        
+                out_sv.write(process_template_sv(template, bfm_name, info))
+                out_c.write(generate_dpi_c(bfm_name, info))
+        
+            out_c.write("#ifdef __cplusplus\n")
+            out_c.write("}\n")
+            out_c.write("#endif\n")
+
+def bfm_generate(args):
+    '''
+    Generate BFM files required for simulation
+    '''
+    
+    if args.o is None:
+        if args.language == "vlog":
+            args.o = "cocotb_bfms.v"
+        elif args.language == "sv":
+            args.o = "cocotb_bfms.sv"
+        elif args.language == "vhdl":
+            args.o = "cocotb_bfms.vhdl"
+            
+    if args.language == "vlog":
+        bfm_generate_vl(args)
+    elif args.language == "sv":
+        bfm_generate_sv(args)
+    elif args.language == "vhdl":
+        raise Exception("VHDL currently unsupported")
+
+    
+# def bfm_list(args):
+#     inst = BfmMgr.inst()
+#     
+#     # TODO: implement list command
+#     
+#     print("inst=" + str(inst))
+# 
+#     print("Number of keys: " + str(len(inst.bfm_type_info_m.keys())))
+#     for t in inst.bfm_type_info_m.keys():
+#         print("BFM: \"" + str(t) + "\"")
```

Comment:
Just a reminder that we should either finish or remove this code before merging

Reply:
Decided to omit this for now.